environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java

```diff
@@ -177,26 +177,31 @@ private EnvironmentDto changeCredential(String accountId, String environmentName
             Environment environment) {
-        ValidationResult validateKey = environmentService.getValidatorService().validateEncryptionKeyUrl(dto.getEncryptionKeyUrl(),
-                accountId);
-        if (!validateKey.hasError()) {
-            AzureParameters azureParameters = (AzureParameters) environment.getParameters();
-            azureParameters.setEncryptionKeyUrl(dto.getEncryptionKeyUrl());
-            azureParameters.setEncryptionKeyResourceGroupName(dto.getEncryptionKeyResourceGroupName());
-            //creating the DES
-            try {
-                CreatedDiskEncryptionSet createdDiskEncryptionSet = environmentEncryptionService.createEncryptionResources(
-                        environmentDtoConverter.environmentToDto(environment));
-                azureParameters.setDiskEncryptionSetId(createdDiskEncryptionSet.getDiskEncryptionSetId());
-                azureParametersRepository.save(azureParameters);
-            } catch (Exception e) {
-                throw new BadRequestException(e);
+        AzureParameters azureParameters = (AzureParameters) environment.getParameters();
+        if (azureParameters.getEncryptionKeyUrl() == null) {
+            ValidationResult validateKey = environmentService.getValidatorService().validateEncryptionKeyUrl(dto.getEncryptionKeyUrl(),
+                    accountId);
+            if (!validateKey.hasError()) {
+                azureParameters.setEncryptionKeyUrl(dto.getEncryptionKeyUrl());
+                azureParameters.setEncryptionKeyResourceGroupName(dto.getEncryptionKeyResourceGroupName());
+                //creating the DES
+                try {
+                    CreatedDiskEncryptionSet createdDiskEncryptionSet = environmentEncryptionService.createEncryptionResources(
+                            environmentDtoConverter.environmentToDto(environment));
+                    azureParameters.setDiskEncryptionSetId(createdDiskEncryptionSet.getDiskEncryptionSetId());
+                    azureParametersRepository.save(azureParameters);
+                } catch (Exception e) {
+                    throw new BadRequestException(e);
+                }
+                LOGGER.debug("Successfully created the Disk encryption set for the environment {}.", environmentName);
+            } else {
+                throw new BadRequestException(validateKey.getFormattedErrors());
             }
-            LOGGER.debug("Successfully created the Disk encryption set for the environment {}.", environmentName);
+            return environmentDtoConverter.environmentToDto(environment);
         } else {
-            throw new BadRequestException(validateKey.getFormattedErrors());
+            throw new BadRequestException(String.format("Encryption Key '%s' is already set for the environment '%s'. " +
+                    "Modifying the encryption key is not allowed.", azureParameters.getEncryptionKeyUrl(), environmentName));
```

Comment:
What if the user specifies exactly the same key in the edit request that was used earlier to encrypt the environment? Is this possible, or rather there is a validation for that scenario in somewhere else? (Strictly speaking, that should not really count as an "edit" as no changes would be made to the encryption settings. Perhaps the request could be just ignored silently in that case, but I leave that to you to decide how to deal with this.)

Reply:
I have updated the code to not throw any error if user tries to update encryption resources using the same encryption key, instead log the info message and return the encryption details as it is.